From 21b9eb23b31928defb5daf909b9a98dba3d32e32 Mon Sep 17 00:00:00 2001
From: Li Jinyue <lijinyue@huawei.com>
Date: Thu, 14 Dec 2017 17:04:54 +0800
Subject: [PATCH 37/60] BACKPORT: futex: Prevent overflow by strengthen input
 validation

UBSAN reports signed integer overflow in kernel/futex.c:

 UBSAN: Undefined behaviour in kernel/futex.c:2041:18
 signed integer overflow:
 0 - -2147483648 cannot be represented in type 'int'

Add a sanity check to catch negative values of nr_wake and nr_requeue.

Signed-off-by: Li Jinyue <lijinyue@huawei.com>
Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Erick Reyes <erickreyes@google.com>
Signed-off-by: Oleg Matcovschi <omatcovschi@google.com>
Cc: peterz@infradead.org
Cc: dvhart@infradead.org
Cc: stable@vger.kernel.org

Link: https://lkml.kernel.org/r/1513242294-31786-1-git-send-email-lijinyue@huawei.com
Cherry-picked from fbe0e839d1e22d88810f3ee3e2f1479be4c0aa4a

Bug: 76106267
Change-Id: I954cc2848678318b60ec3f103d0c15f87b4605a4
CVE-2018-6927
Signed-off-by: Kevin F. Haggerty <haggertk@lineageos.org>
---
 kernel/futex.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/futex.c b/kernel/futex.c
index 38b4e4083a4..7de0172ce42 100644
--- a/kernel/futex.c
+++ b/kernel/futex.c
@@ -1463,6 +1463,9 @@ static int futex_requeue(u32 __user *uaddr1, unsigned int flags,
 	struct plist_head *head1;
 	struct futex_q *this, *next;
 
+	if (nr_wake < 0 || nr_requeue < 0)
+		return -EINVAL;
+
 	if (requeue_pi) {
 		/*
 		 * Requeue PI only works on two distinct uaddrs. This
-- 
2.17.1

